# Assignment 1

Ensure you have referred to the installation document and installed the following for you to take through the assignment and projects. Check the boxes

- [x] conda
- [x] JupyterBook
- [x] psql command-line interface
- [x] pgadmin
- [x] Setup your AWS account
- [x] Created your postgres RDS instance (you can use the instance that you created in worksheet1)
- [x] Made credentials.json file in your notebook folder with the connection details

## Mechanics
***Must be submitted as `.ipynb`, `.html`, and any other screenshot images (Please don't upload a folder or a zipped folder).***

(5 points)

- [x] Upload `.ipynb`, `.html`, and any other files asked for to canvas. (you should upload both `.ipynb` and `.html` file.)
- [x] Screenshots and code output rendered correctly.
- [x] Credentials not exposed in the notebook.

You need to make sure that you installed all necessary packages that are needed for this assignment. Please refer to installation instructions conda on more information on necessary packages.

## Case Study
Twitter is a massive platform.  There are 300+ million users on Twitter, and it is a source of information for current events, social movements and, financial information.  It has been shown in a number of cases that information from Twitter can mobilize a large number of individuals.  From #blacklivesmatter to other forms of *hashtag* activism, social media can play an important role in informing and mobilizing individuals.

This same activity can be extended to financial information.  The introduction of "cashtags" to twitter has allowed individuals to connect and discuss stocks, but it has also given stock promoters a method for promoting low value stocks, to "pump and dump".  Some researchers have analyzed the use of cashtags on Twitter.  We will use a similar method to look at the data, but we will ask a slightly different question.

### Reading
Hentschel M, Alonso O. 2014. Follow the money: A study of cashtags on Twitter. *First Monday*. URL: https://firstmonday.org/ojs/index.php/fm/article/view/5385/4109

#### Supplementary Information

* Evans, L., Owda, M., Crockett, K., & Vilas, A. F. (2019). A methodology for the resolution of cashtag collisions on Twitter–A natural language processing & data fusion approach. *Expert Systems with Applications*, **127**, 353-369.
* Evans, L., Owda, M., Crockett, K., & Vilas, A. F. (2021). [Credibility assessment of financial stock tweets](https://www.sciencedirect.com/science/article/pii/S0957417420310356). *Expert Systems with Applications*, **168**, 114351.
* Cresci, S., Lillo, F., Regoli, D., Tardelli, S., & Tesconi, M. (2019). Cashtag Piggybacking: Uncovering Spam and Bot Activity in Stock Microblogs on Twitter. *ACM Transactions on the Web (TWEB)*, **13(2)**, 11.

#### Raw Data source
I document the source of ticker data below.  The tweet data we use here comes from a dataset used in Cresci *et al* (2019) referenced above.  The data is available through Zenodo using the dataset's DOI: [10.5281/zenodo.2686861](https://doi.org/10.5281/zenodo.2686861). 

This is for your reference. I have already created the schema, tables, normalized data and loaded it to the database.

## Data dictionary

This is information to get to know more on the twitter data that you will be working on.

### Twitter CashTag Records

  * Title: Cashtag Piggybacking dataset - Twitter dataset enriched with financial data.
  * URI: https://doi.org/10.5281/zenodo.2686861
  * Keywords: cashtag, tweets, twitter, piggybacking, social, bot
  * Publication Date: May 9, 2019
  * Publisher: Zenodo
  * Creator: Cresci, Stefano; Lillo, Fabrizio; Regoli, Daniele;  Tardelli, Serena; Tesconi, Maurizio
  * Contact Point: N/A
  * Spatial Coverage: Global
  * Temporal Coverage: May - September 2017
  * Language: Multilingual
  * Date & Time Formats: "Thu May 18 22:00:00 +0000 2017"
  * Data Version: 1.0
  * Access Date: March 2, 2021

### NYSE Stock Symbol Dataset

  * Title: NYSE Symbol Directory
  * URI: ftp://ftp.nasdaqtrader.com/SymbolDirectory/otherlisted.txt
  * Keywords: stock symbols, nyse
  * Publication Date: May 2, 2021
  * Publisher: nasdaqtrader.com
  * Creator: N/A
  * Contact Point: N/A
  * Spatial Coverage: Global
  * Temporal Coverage: N/A
  * Language: English
  * Date & Time Formats: N/A
  * Data Version: N/A
  * Access Date: March 2, 2021

### NYSE Stock Trading Data 

  * Title: Yahoo! Finance Data Daily Stock Price
  * URI: https://finance.yahoo.com/
  * Keywords: stock symbols, nyse, stock price
  * Publication Date: N/A
  * Publisher: Yahoo! Finance
  * Creator: N/A
  * Contact Point: N/A
  * Spatial Coverage: New York, New York, USA
  * Temporal Coverage: May - September 2017
  * Language: English
  * Date & Time Formats: 2003-02-19
  * Data Version: N/A
  * Access Date: March 2, 2021
  
  ## Data Model

For each stock symbol, by day, we want the count of tweets that mention the symbol, the mean stock value on that day, and the name of the stock.

Yahoo! Finance data is generated only for NYSE stocks, so we will ignore other stock symbols in the dataset.

### NYSE Security Names

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| symbolid | integer/serial | Unique numeric identifier | Unique integer |
| nasdaqsymbol | text | NYSE Symbol Directory | Security abbreviation | Unique, alphanumeric, upper case, including (.-=+) |
| securityname | text | NYSE Symbol Directory | Security name | Alphanumeric, unique   |

### NYSE Stock Values

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| symbolid | int | References securities | Integer link to the securities table. | FOREIGN KEY |
| date | date | Yahoo! Finance | Date of stock values | Valid datetime (yyyy-MM-DD) |
| volume | numeric |  Yahoo! Finance | volume of stock traded on date | Positive integer value |
| open | float |  Yahoo! Finance | USD value of security at trading open | Positive float value |
| close | float |  Yahoo! Finance | USD value of security at trading close | Positive float value |

### Twitter Users

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| userid | bigint | Twitter API | Unique numeric integer to identify a user | Positive integer |
| username | text | Twitter API | Optional user name (not used currently) | Text string with valid twitter handle |

### Tweets

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| tweetid | bigint | Twitter API | Unique numeric ID for tweets | Positive integer |
| userid | bigint | Twitter API | Unique numeric ID for users | FOREIGN KEY |
| text | text | Twitter API | Text content of a Tweet | Valid text for tweet content |
| createddate | datetime | Twitter API | Date with timestamp | Valid Datetime between May and September 2017 |

### Retweets

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| userid | bigint | Twitter API | User ID from above | FOREIGN KEY |
| retweetid | bigint | Twitter API | Tweet ID for retweet | Integer, unique |
| tweetid | bigint | Twitter API | Tweet ID for retweet | FOREIGN KEY |
| createddate | datetime | Twitter API | Date with timestamp | Valid Datetime between May and September 2017 |

### Replies

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| tweetid | bigint | Twitter API | Tweet ID the user is replying to | FOREIGN KEY |
| replyid | bigint | Twitter API | Tweet ID for reply | FOREIGN KEY |

### Cashtags

| column | format | source | description | Validation |
| ------ | ------ | ---------- | ------------ | ------------- |
| tweetid | bigint | Twitter API | Tweet ID the user is replying to | FOREIGN KEY |
| symbolid | integer | securities | Cashtag Symbol | FOREIGN KEY |


## From Tables to Tables:

```
CREATE TABLE IF NOT EXISTS symbols (
  symbolid serial PRIMARY KEY,
  nasdaqsymbol varchar NOT NULL,
  securityname text NOT NULL
  CONSTRAINT valid_symbol CHECK (nasdaqsymbol ~* '^[A-Z\.\=\-\+\#\^]+$')
);

CREATE TABLE IF NOT EXISTS stockvalues (
    symbolid int REFERENCES symbols(symbolid),
    date date CHECK (date >= '2017-05-01' AND date < '2017-10-01'),
    open float CHECK (open > 0),
    close float CHECK (close > 0)
);

CREATE TABLE IF NOT EXISTS users (
    userid bigint PRIMARY KEY,
    username text
);

CREATE TABLE IF NOT EXISTS tweets (
    tweetid bigint PRIMARY KEY,
    userid bigint REFERENCES users(userid),
    tweet text NOT NULL,
    createddate datetime NOT NULL
);

CREATE TABLE IF NOT EXISTS retweets (
    userid bigint REFERENCES users(userid),
    retweetid bigint NOT NULL PRIMARY KEY,
    tweetid bigint REFERENCES tweets(tweetid),
    createddate datetime NOT NULL
);

CREATE TABLE IF NOT EXISTS replies (
    tweetid bigint REFERENCES tweets(tweetid),
    replyid bigint REFERENCES tweets(tweetid) UNIQUE PRIMARY KEY
);

CREATE TABLE IF NOT EXISTS cashtags (
    tweetid bigint REFERENCES tweets(tweetid),
    symbolid int REFERENCES symbols(symbolid)
);
```

<img src="ERDiagram.png">

With this model, we can start to put together queries. Now its time to get to questions in the milestone.

Let's get started;

## Imports and configurations
You can use any other packages if you want to do some visualization; But please be aware that you want to install those packages using `conda` or `pip`.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import psycopg2

%matplotlib inline
%load_ext sql
%config SqlMagic.displaylimit = 20
%config SqlMagic.autolimit = 30

Before running the following cell, make sure that you have the correct login information in the `credentials.json` file:

In [3]:
import json
import urllib.parse

with open('credentials.json') as f:
    login = json.load(f)
    
username = login['user']
password = urllib.parse.quote(login['password'])
host = login['host']
port = login['port']

## Questions

***1.  Identify elements of the potential dataset(s) that match each of the four Vs of Big Data: (Please edit here with your answer)***

rubric={reasoning:10}

  a.  Velocity: The dataset tweets in the database postgres usually runs for a seconds to large value of milliseconds. Since, we have hard coded the data into the server so it is not changing otherwise, if the data is collected directly from the source we would be able to obtain real time data.

  b.  Veracity: The majority data has been collected from twitter data source which is trustworthy and that the tweets have actually be made. Another, source is Yahoo finance which is very trustworthy as it report directly from the stock market which is affected by the buy and sell activities of stocks. 

  c.  Volume: The volume of the database is 5425 MB. Within the database postgres there are various datasets such as fake data and tweets. Each dataset has several tables with multiple rows. If the data is real time then the size would be more depending on availability on the server.

  d.  Variety: The dataset tweets have been collected from Yahoo finance, NYSE symbol directory ,securities and twitter. The data is normalized with well defined tables and units. Since, the data is collected from an internet source the chances of error would be realtively low.

***2. Load the dumps([tweets.sql](https://canvas.ubc.ca/files/18519971/download?download_frd=1)). As the file size was big, I had to zip it before uploading. So make sure you unzip it before loading. You can paste the commands what you used in your terminal here. You should mask/remove your hostname for security reasons.***

rubric={correctness:5}


---Below this line paste the commands you used in terminal to load the dump to your database---


#### \i /Users/yashika/Downloads/tweets.sql

***3. Just to test your connection, your first SQL query on your cloud database instance (RDS). Using a SQL query, return the number of rows in this table `tweets` in `tweets` schema.***

rubric={correctness:15}

Use the following cell (and copy it as needed) to connect to the database that you need for a question. Remember that if you're using the same database for a few questions, you don't need to reconnect each time. Only use the following cell for establishing the first connection, and for when you want to switch from one to another.

> Here the postgres is the default database name that you chose while setting up RDS (check instructions). In the question above the dumps are used to load the data to this default database

In [3]:
%sql postgresql://{username}:{password}@{host}:{port}/postgres

'Connected: postgres@postgres'

In [16]:
%%sql
SELECT COUNT (*) AS "row_count"
FROM tweets.tweets;

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
1 rows affected.


[(200,)]

***4. Ask a question you want to answer on Twitter (the data you loaded in Question 2). Based on your question, create a mini-warehouse and proper indexing for the best performance for your query. Finally, explain the decisions you make.***

Following is the breakdown of how you need to proceed.

- Question identification
- Write your reasoning on how your question can be solved? SQL query that you are planning to use.
- Build your mini warehouse and reasoning on the decisions you make
- Add indexing and reasoning to the decisions you make
- Your SQL query on your mini warehouse after all indexing
- Print/visualize your answer.

rubric={interesting/correctness/reasoning:20 }

The question you are asking can't be simple (but at the same time doesn't have to be too complicated). You can check the example we discussed in lecture 4, but please note it shouldn't be of the same theme. 

#### Question: Does the number of tweets made about the top ten stock affect the overall fluctuation in their closing values?

###### To answer this question, I will use the query mentioned in the following cell which I am creating by joining the tables and then using a subquery in the subsequent to find the last day on which the tweets were made, after this I intend to use aggregate functions to find the fluctuations in closing values of top ten stock and count the tweets made for those stocks.

In [48]:
%%time
%%sql

EXPLAIN ANALYZE
SELECT ct.tweetid, ct.symbolid, sy.securityname, sv.close
FROM tweets.cashtags AS ct
INNER JOIN tweets.cleantweets AS cl ON ct.tweetid = cl.tweetid
INNER JOIN tweets.symbols AS sy ON ct.symbolid = sy.symbolid
INNER JOIN tweets.stockvalues AS sv ON sv.symbolid = sy.symbolid AND sv.date = date_trunc('day', cl.createdate)::date

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
30 rows affected.
CPU times: user 4.27 ms, sys: 2.16 ms, total: 6.43 ms
Wall time: 13.1 s


[('Gather  (cost=632102.14..806559.60 rows=1368257 width=74) (actual time=9006.128..12721.881 rows=3539052 loops=1)',),
 ('  Workers Planned: 2',),
 ('  Workers Launched: 2',),
 ('  ->  Merge Join  (cost=631102.14..668733.90 rows=570107 width=74) (actual time=9139.384..11004.773 rows=1179684 loops=3)',),
 ("        Merge Cond: ((ct.symbolid = sy.symbolid) AND (((date_trunc('day'::text, cl.createdate))::date) = sv.date))",),
 ('        ->  Sort  (cost=573996.88..578775.82 rows=1911579 width=20) (actual time=8398.684..9006.133 rows=1529263 loops=3)',),
 ("              Sort Key: ct.symbolid, ((date_trunc('day'::text, cl.createdate))::date)",),
 ('              Sort Method: external merge  Disk: 54880kB',),
 ('              Worker 0:  Sort Method: external merge  Disk: 55296kB',),
 ('              Worker 1:  Sort Method: external merge  Disk: 60576kB',),
 ('              ->  Parallel Hash Join  (cost=214437.97..296151.65 rows=1911579 width=20) (actual time=3729.302..5832.054 rows=1529263 loops=3)',),
 ('                    Hash Cond: (ct.tweetid = cl.tweetid)',),
 ('                    ->  Parallel Seq Scan on cashtags ct  (cost=0.00..43914.79 rows=1911579 width=12) (actual time=0.013..255.265 rows=1529263 loops=3)',),
 ('                    ->  Parallel Hash  (cost=164197.21..164197.21 rows=2890221 width=16) (actual time=2875.277..2875.278 rows=2312556 loops=3)',),
 ('                          Buckets: 131072  Batches: 128  Memory Usage: 3616kB',),
 ('                          ->  Parallel Seq Scan on cleantweets cl  (cost=0.00..164197.21 rows=2890221 width=16) (actual time=0.043..1333.697 rows=2312556 loops=3)',),
 ('        ->  Materialize  (cost=57103.69..58929.21 rows=365104 width=74) (actual time=740.682..1061.052 rows=1400567 loops=3)',),
 ('              ->  Sort  (cost=57103.69..58016.45 rows=365104 width=74) (actual time=740.678..859.840 rows=365092 loops=3)',),
 ('                    Sort Key: sy.symbolid, sv.date',),
 ('                    Sort Method: external merge  Disk: 27320kB',),
 ('                    Worker 0:  Sort Method: external merge  Disk: 27320kB',),
 ('                    Worker 1:  Sort Method: external merge  Disk: 27320kB',),
 ('                    ->  Hash Join  (cost=209.72..7145.83 rows=365104 width=74) (actual time=5.491..271.152 rows=365104 loops=3)',),
 ('                          Hash Cond: (sv.symbolid = sy.symbolid)',),
 ('                          ->  Seq Scan on stockvalues sv  (cost=0.00..5977.04 rows=365104 width=16) (actual time=0.033..74.595 rows=365104 loops=3)',),
 ('                          ->  Hash  (cost=133.21..133.21 rows=6121 width=58) (actual time=5.403..5.404 rows=6121 loops=3)',),
 ('                                Buckets: 8192  Batches: 1  Memory Usage: 613kB',),
 ('                                ->  Seq Scan on symbols sy  (cost=0.00..133.21 rows=6121 width=58) (actual time=0.018..2.602 rows=6121 loops=3)',),
 ('Planning Time: 2.919 ms',),
 ('Execution Time: 12987.583 ms',)]

####  When we run the query originally it took about 13.1 seconds.

In [4]:
%%sql

SELECT symbolid
FROM tweets.stockvalues
WHERE date = (SELECT MAX(date) FROM tweets.stockvalues)
ORDER BY close DESC
LIMIT 10;

Environment variable $DATABASE_URL not set, and no connect string given.
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


In [5]:
%%time
%%sql

CREATE MATERIALIZED VIEW if not exists tweets.materialized AS
SELECT ct.tweetid, ct.symbolid, sy.securityname, sv.close
FROM tweets.cashtags AS ct
INNER JOIN tweets.cleantweets AS cl ON ct.tweetid = cl.tweetid
INNER JOIN tweets.symbols AS sy ON ct.symbolid = sy.symbolid
INNER JOIN tweets.stockvalues AS sv ON sv.symbolid = sy.symbolid AND sv.date = date_trunc('day', cl.createdate)::date

Environment variable $DATABASE_URL not set, and no connect string given.
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])
CPU times: user 372 µs, sys: 160 µs, total: 532 µs
Wall time: 482 µs


#### After creating a materialized view the query ran in 11.9 s. To find the stocks made on the last day further reduced the running time to 1.6s. The materialized view pre computes the data and takes less time to perform joins

In [54]:
%%time
%%sql

EXPLAIN ANALYZE 
SELECT * FROM tweets.materialized
WHERE symbolid IN (SELECT symbolid
FROM tweets.stockvalues
WHERE date = (SELECT MAX(date) FROM tweets.stockvalues)
ORDER BY close DESC
LIMIT 10);

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
27 rows affected.
CPU times: user 3.39 ms, sys: 1.94 ms, total: 5.33 ms
Wall time: 1.6 s


[('Hash Semi Join  (cost=12066.05..95880.24 rows=10119 width=58) (actual time=561.365..1561.909 rows=13824 loops=1)',),
 ('  Hash Cond: (materialized.symbolid = "ANY_subquery".symbolid)',),
 ('  ->  Seq Scan on materialized  (cost=0.00..74410.13 rows=3539613 width=58) (actual time=0.007..514.971 rows=3539052 loops=1)',),
 ('  ->  Hash  (cost=12065.93..12065.93 rows=10 width=4) (actual time=361.499..361.723 rows=10 loops=1)',),
 ('        Buckets: 1024  Batches: 1  Memory Usage: 9kB',),
 ('        ->  Subquery Scan on "ANY_subquery"  (cost=12064.68..12065.93 rows=10 width=4) (actual time=361.484..361.714 rows=10 loops=1)',),
 ('              ->  Limit  (cost=12064.68..12065.83 rows=10 width=12) (actual time=361.483..361.711 rows=10 loops=1)',),
 ('                    InitPlan 1 (returns $1)',),
 ('                      ->  Finalize Aggregate  (cost=6010.70..6010.71 rows=1 width=4) (actual time=257.374..257.455 rows=1 loops=1)',),
 ('                            ->  Gather  (cost=6010.59..6010.70 rows=1 width=4) (actual time=253.606..257.446 rows=2 loops=1)',),
 ('                                  Workers Planned: 1',),
 ('                                  Workers Launched: 1',),
 ('                                  ->  Partial Aggregate  (cost=5010.59..5010.60 rows=1 width=4) (actual time=245.335..245.336 rows=1 loops=2)',),
 ('                                        ->  Parallel Seq Scan on stockvalues stockvalues_1  (cost=0.00..4473.67 rows=214767 width=4) (actual time=0.007..85.416 rows=182552 loops=2)',),
 ('                    ->  Gather Merge  (cost=6053.97..6284.77 rows=2007 width=12) (actual time=361.482..361.626 rows=10 loops=1)',),
 ('                          Workers Planned: 1',),
 ('                          Params Evaluated: $1',),
 ('                          Workers Launched: 1',),
 ('                          ->  Sort  (cost=5053.96..5058.98 rows=2007 width=12) (actual time=96.713..96.715 rows=7 loops=2)',),
 ('                                Sort Key: stockvalues.close DESC',),
 ('                                Sort Method: top-N heapsort  Memory: 25kB',),
 ('                                Worker 0:  Sort Method: top-N heapsort  Memory: 25kB',),
 ('                                ->  Parallel Seq Scan on stockvalues  (cost=0.00..5010.59 rows=2007 width=12) (actual time=0.020..87.976 rows=1728 loops=2)',),
 ('                                      Filter: (date = $1)',),
 ('                                      Rows Removed by Filter: 180824',),
 ('Planning Time: 0.254 ms',),
 ('Execution Time: 1562.513 ms',)]

#### The materialized view on the last possible date in the data set took about 1.6s to execute as it has alsready performed the joins and thus, retrieving the data for max date took even lesser time to execute.

In [55]:
%%sql
CREATE INDEX if not exists symbolid_index ON tweets.materialized(symbolid);

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
Done.


[]

In [56]:
%%time
%%sql

EXPLAIN ANALYZE 
SELECT * FROM tweets.materialized
WHERE symbolid IN (SELECT symbolid
FROM tweets.stockvalues
WHERE date = (SELECT MAX(date) FROM tweets.stockvalues)
ORDER BY close DESC
LIMIT 10);

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
27 rows affected.
CPU times: user 3.91 ms, sys: 2.07 ms, total: 5.98 ms
Wall time: 271 ms


[('Nested Loop  (cost=12066.38..26047.14 rows=10118 width=58) (actual time=215.611..229.702 rows=13824 loops=1)',),
 ('  ->  HashAggregate  (cost=12065.95..12066.05 rows=10 width=4) (actual time=215.447..215.665 rows=10 loops=1)',),
 ('        Group Key: stockvalues.symbolid',),
 ('        Batches: 1  Memory Usage: 24kB',),
 ('        ->  Limit  (cost=12064.68..12065.83 rows=10 width=12) (actual time=215.122..215.329 rows=10 loops=1)',),
 ('              InitPlan 1 (returns $1)',),
 ('                ->  Finalize Aggregate  (cost=6010.70..6010.71 rows=1 width=4) (actual time=142.376..142.441 rows=1 loops=1)',),
 ('                      ->  Gather  (cost=6010.59..6010.70 rows=1 width=4) (actual time=142.250..142.433 rows=2 loops=1)',),
 ('                            Workers Planned: 1',),
 ('                            Workers Launched: 1',),
 ('                            ->  Partial Aggregate  (cost=5010.59..5010.60 rows=1 width=4) (actual time=134.627..134.628 rows=1 loops=2)',),
 ('                                  ->  Parallel Seq Scan on stockvalues stockvalues_1  (cost=0.00..4473.67 rows=214767 width=4) (actual time=0.008..63.633 rows=182552 loops=2)',),
 ('              ->  Gather Merge  (cost=6053.97..6284.77 rows=2007 width=12) (actual time=215.120..215.260 rows=10 loops=1)',),
 ('                    Workers Planned: 1',),
 ('                    Params Evaluated: $1',),
 ('                    Workers Launched: 1',),
 ('                    ->  Sort  (cost=5053.96..5058.98 rows=2007 width=12) (actual time=61.642..61.644 rows=8 loops=2)',),
 ('                          Sort Key: stockvalues.close DESC',),
 ('                          Sort Method: top-N heapsort  Memory: 25kB',),
 ('                          Worker 0:  Sort Method: top-N heapsort  Memory: 25kB',),
 ('                          ->  Parallel Seq Scan on stockvalues  (cost=0.00..5010.59 rows=2007 width=12) (actual time=0.016..60.575 rows=1728 loops=2)',),
 ('                                Filter: (date = $1)',),
 ('                                Rows Removed by Filter: 180824',),
 ('  ->  Index Scan using symbolid_index on materialized  (cost=0.43..1382.53 rows=1558 width=58) (actual time=0.098..1.026 rows=1382 loops=10)',),
 ('        Index Cond: (symbolid = stockvalues.symbolid)',),
 ('Planning Time: 1.294 ms',),
 ('Execution Time: 231.181 ms',)]

#### After this, I created a b-tree index on symbolid and executed the materialized view query for tweets made on the last date and the execution time further reduced to 271 ms.

In [57]:
%%time
%%sql

EXPLAIN ANALYZE 
SELECT symbolid, securityname, COUNT(tweetid), MAX(close)-MIN(close) "close_value_change" 
FROM tweets.materialized
WHERE symbolid IN (SELECT symbolid
FROM tweets.stockvalues
WHERE date = (SELECT MAX(date) FROM tweets.stockvalues)
ORDER BY close DESC
LIMIT 10)
GROUP BY symbolid, securityname;

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
30 rows affected.
CPU times: user 3.84 ms, sys: 1.83 ms, total: 5.67 ms
Wall time: 469 ms


[('HashAggregate  (cost=26173.61..26300.09 rows=10118 width=58) (actual time=332.429..333.875 rows=10 loops=1)',),
 ('  Group Key: materialized.symbolid, materialized.securityname',),
 ('  Batches: 1  Memory Usage: 409kB',),
 ('  ->  Nested Loop  (cost=12066.38..26047.14 rows=10118 width=58) (actual time=305.342..320.382 rows=13824 loops=1)',),
 ('        ->  HashAggregate  (cost=12065.95..12066.05 rows=10 width=4) (actual time=305.310..306.708 rows=10 loops=1)',),
 ('              Group Key: stockvalues.symbolid',),
 ('              Batches: 1  Memory Usage: 24kB',),
 ('              ->  Limit  (cost=12064.68..12065.83 rows=10 width=12) (actual time=305.294..306.680 rows=10 loops=1)',),
 ('                    InitPlan 1 (returns $1)',),
 ('                      ->  Finalize Aggregate  (cost=6010.70..6010.71 rows=1 width=4) (actual time=122.521..122.584 rows=1 loops=1)',),
 ('                            ->  Gather  (cost=6010.59..6010.70 rows=1 width=4) (actual time=121.744..122.574 rows=2 loops=1)',),
 ('                                  Workers Planned: 1',),
 ('                                  Workers Launched: 1',),
 ('                                  ->  Partial Aggregate  (cost=5010.59..5010.60 rows=1 width=4) (actual time=117.531..117.532 rows=1 loops=2)',),
 ('                                        ->  Parallel Seq Scan on stockvalues stockvalues_1  (cost=0.00..4473.67 rows=214767 width=4) (actual time=0.007..62.487 rows=182552 loops=2)',),
 ('                    ->  Gather Merge  (cost=6053.97..6284.77 rows=2007 width=12) (actual time=305.293..306.614 rows=10 loops=1)',),
 ('                          Workers Planned: 1',),
 ('                          Params Evaluated: $1',),
 ('                          Workers Launched: 1',),
 ('                          ->  Sort  (cost=5053.96..5058.98 rows=2007 width=12) (actual time=168.368..168.370 rows=10 loops=2)',),
 ('                                Sort Key: stockvalues.close DESC',),
 ('                                Sort Method: top-N heapsort  Memory: 25kB',),
 ('                                Worker 0:  Sort Method: top-N heapsort  Memory: 25kB',),
 ('                                ->  Parallel Seq Scan on stockvalues  (cost=0.00..5010.59 rows=2007 width=12) (actual time=0.017..167.838 rows=1728 loops=2)',),
 ('                                      Filter: (date = $1)',),
 ('                                      Rows Removed by Filter: 180824',),
 ('        ->  Index Scan using symbolid_index on materialized  (cost=0.43..1382.53 rows=1558 width=58) (actual time=0.014..0.687 rows=1382 loops=10)',),
 ('              Index Cond: (symbolid = stockvalues.symbolid)',),
 ('Planning Time: 0.246 ms',),
 ('Execution Time: 334.016 ms',)]

#### Now, to find the top ten closing value changes of stocks the indexing exectution time is 469 ms. This is because of the aggregate functions applied on the columns to return the answer to the question originally asked.

In [59]:
%%time
%%sql

SELECT symbolid, securityname, COUNT(tweetid), MAX(close)-MIN(close) "close_value_change" 
FROM tweets.materialized
WHERE symbolid IN (SELECT symbolid
FROM tweets.stockvalues
WHERE date = (SELECT MAX(date) FROM tweets.stockvalues)
ORDER BY close DESC
LIMIT 10)
GROUP BY symbolid, securityname
ORDER BY close_value_change DESC;

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
10 rows affected.
CPU times: user 3.47 ms, sys: 1.61 ms, total: 5.08 ms
Wall time: 120 ms


[(898, 'Camber Energy, Inc. Common Stock', 1495, 179687.5),
 (2500, 'Direxion Daily S&P Oil & Gas Exp. & Prod. Bull 2X Shares', 236, 5700.16357421875),
 (5667, 'ProShares Trust Ultra VIX Short Term Futures ETF', 6259, 1686.0),
 (3305, 'Direxion Daily S&P Biotech Bear 3X Shares', 717, 1166.1193237304688),
 (5049, 'Direxion Daily Semiconductor Bear 3x Shares', 154, 1031.107666015625),
 (3066, 'Direxion Daily Junior Gold Miners Index Bear 2X Shares', 1990, 990.9801635742188),
 (742, 'Direxion Daily Brazil Bull 2X Shares', 298, 860.5069580078125),
 (4875, 'Seaboard Corporation Common Stock', 147, 758.002197265625),
 (3967, 'NVR, Inc. Common Stock', 1314, 556.97998046875),
 (3584, 'Markel Corporation Common Stock', 1214, 134.0599365234375)]

#### After indexing and executing the query to find the top ten highest closing value change of the stock and the number of tweets made on the original query is returned in 120 ms. The materialized view and indexing helped in reducing the overall time of executing the query and the plot shows count of tweets vs the close_value_change to see if there is any relationship between the two. 

In [68]:
import plotly

In [79]:
%%sql query_output <<

SELECT symbolid, securityname, COUNT(tweetid), MAX(close)-MIN(close) "close_value_change" 
FROM tweets.materialized
WHERE symbolid IN (SELECT symbolid
FROM tweets.stockvalues
WHERE date = (SELECT MAX(date) FROM tweets.stockvalues)
ORDER BY close DESC
LIMIT 10)
GROUP BY symbolid, securityname
ORDER BY close_value_change DESC;

 * postgresql://postgres:***@yashikadatabase1.c2uqp9nogr60.us-west-2.rds.amazonaws.com:5432/postgres
10 rows affected.
Returning data to local variable query_output


In [83]:
import plotly.express as px
import pandas as pd
fig = px.scatter(query_output, 
              x='count',
              y='close_value_change',
              title='Overall fluctuations in close values by tweet count.', 
              hover_name='symbolid')
fig.show()

##### The plot does not show any relationship betweent the fluctuations in closing value of top ten stocks in relation to the number of tweets tweeted on it on the last day of the dataset.